src/librustdoc/clean/utils.rs

```diff
@@ -427,39 +427,46 @@ crate fn get_auto_trait_and_blanket_impls(
-        Res::Def(DefKind::Fn, i) => (i, TypeKind::Function),
-        Res::Def(DefKind::TyAlias, i) => (i, TypeKind::Typedef),
-        Res::Def(DefKind::Enum, i) => (i, TypeKind::Enum),
-        Res::Def(DefKind::Trait, i) => (i, TypeKind::Trait),
+crate fn register_res_inner(tcx: TyCtxt<'_>, res: Res) -> (DefId, Option<TypeKind>) {
+    debug!("register_res_inner({:?})", res);
+
+    match res {
+        Res::Def(DefKind::Fn, i) => (i, Some(TypeKind::Function)),
+        Res::Def(DefKind::TyAlias, i) => (i, Some(TypeKind::Typedef)),
+        Res::Def(DefKind::Enum, i) => (i, Some(TypeKind::Enum)),
+        Res::Def(DefKind::Trait, i) => (i, Some(TypeKind::Trait)),
         Res::Def(DefKind::AssocTy | DefKind::AssocFn | DefKind::AssocConst, i) => {
-            (cx.tcx.parent(i).unwrap(), TypeKind::Trait)
+            (tcx.parent(i).expect("cannot get parent def id"), Some(TypeKind::Trait))
         }
-        Res::Def(DefKind::Struct, i) => (i, TypeKind::Struct),
-        Res::Def(DefKind::Union, i) => (i, TypeKind::Union),
-        Res::Def(DefKind::Mod, i) => (i, TypeKind::Module),
-        Res::Def(DefKind::ForeignTy, i) => (i, TypeKind::Foreign),
-        Res::Def(DefKind::Const, i) => (i, TypeKind::Const),
-        Res::Def(DefKind::Static, i) => (i, TypeKind::Static),
+        Res::Def(DefKind::Struct, i) => (i, Some(TypeKind::Struct)),
+        Res::Def(DefKind::Union, i) => (i, Some(TypeKind::Union)),
+        Res::Def(DefKind::Mod, i) => (i, Some(TypeKind::Module)),
+        Res::Def(DefKind::ForeignTy, i) => (i, Some(TypeKind::Foreign)),
+        Res::Def(DefKind::Const, i) => (i, Some(TypeKind::Const)),
+        Res::Def(DefKind::Static, i) => (i, Some(TypeKind::Static)),
         Res::Def(DefKind::Variant, i) => {
-            (cx.tcx.parent(i).expect("cannot get parent def id"), TypeKind::Enum)
+            (tcx.parent(i).expect("cannot get parent def id"), Some(TypeKind::Enum))
         }
         Res::Def(DefKind::Macro(mac_kind), i) => match mac_kind {
-            MacroKind::Bang => (i, TypeKind::Macro),
-            MacroKind::Attr => (i, TypeKind::Attr),
-            MacroKind::Derive => (i, TypeKind::Derive),
+            MacroKind::Bang => (i, Some(TypeKind::Macro)),
+            MacroKind::Attr => (i, Some(TypeKind::Attr)),
+            MacroKind::Derive => (i, Some(TypeKind::Derive)),
         },
-        Res::Def(DefKind::TraitAlias, i) => (i, TypeKind::TraitAlias),
-        Res::SelfTy(Some(def_id), _) => (def_id, TypeKind::Trait),
-        Res::SelfTy(_, Some((impl_def_id, _))) => return impl_def_id,
-        _ => return res.def_id(),
-    };
-    if did.is_local() {
+        Res::Def(DefKind::TraitAlias, i) => (i, Some(TypeKind::TraitAlias)),
+        Res::SelfTy(Some(def_id), _) => (def_id, Some(TypeKind::Trait)),
+        Res::SelfTy(_, Some((impl_def_id, _))) => (impl_def_id, None),
+        _ => (res.def_id(), None),
+    }
+}
+
+crate fn register_res(cx: &mut DocContext<'_>, res: Res) -> DefId {
```

Comment:
Just wondering, what does `res` means?

Reply:
Resolution. :)

Reply:
It's in the docs: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/def/enum.Res.html